uvicorn/protocols/http/h11_impl.py

```diff
@@ -155,6 +155,21 @@ def _unset_keepalive_if_required(self) -> None:
             self.timeout_keep_alive_task.cancel()
             self.timeout_keep_alive_task = None
 
+    def _should_upgrade(self) -> bool:
+        connection = []
+        upgrade = None
+        for name, value in self.headers:
+            if name == b"connection":
+                connection = [token.lower().strip() for token in value.split(b",")]
+            if name == b"upgrade":
+                upgrade = value.lower()
+
+        return (
+            b"upgrade" in connection
+            and upgrade == b"websocket"
+            and not self.config.ws_ignore_upgrade
```

Comment:
Hm I think we could eliminate the new flag by changing it to this
```suggestion
            and self.ws_protocol_class is not None
```
and removing the corresponding block from `handle_websocket_upgrade`. I think that would achieve the same thing.

Reply:
I think the warning should stay tho 🤔

Reply:
So trying to understand what that block tries to achieve is 
- logging a warning
- logging another warning if no protocol was found
- responding with 400

But should these warnings really be logged on every upgrade request again? Especially the second one.
In my use case this will lead to a lot of unnecessary log messages.

Reply:
What do you think about moving the second log statement to the run method in main.py? I feel like this indicates a possible configuration / setup issue which would be easier to catch if printed up front on startup. We should rephrase to account for the fact that there are applications which will use the default setting `auto` without the need for an implementation to be present.

I am still undecided on what to do with the first warning. Since I already have a custom log handler, I could filter this message out to prevent the broken window for me.

Alternatively we could document the behaviour and necessary remediation inside the docs to help confused devs figure out why their request gets ignored and or turn down the log level.